MockDataGenerator + DB Connection

nicolasantolini opened the pull request:
#### PR Classification
Creation of Mock Data Generator and Verified and Test connection to database.

#### PR Summary
This pull request introduces new services for managing attendance and generates mock data for testing. It also upgrades several packages to improve functionality and maintainability.
- `Program.cs`: Added `UserService`, `CourseService`, and `AttendanceService`, along with a `MockDataGenerator` for testing.
- `CoursesContext.cs`: Updated constructor to use `DbContextOptions<CoursesContext>` and improved the `OnModelCreating` method.
- `Session.cs`: Changed `SetAttendance` method to be asynchronous for better performance.
- New migration files (`20250226101123_InitialCreate.Designer.cs` and `20250226101123_InitialCreate.cs`) were created to reflect the updated database schema.
- `AttendanceSystem.Api.csproj`: Included `local.settings.json` and upgraded several package references, including `Microsoft.ApplicationInsights` and `Microsoft.EntityFrameworkCore`.

nicolasantolini committed:
Update database configuration and add mock data generator

- Added new using directives in Program.cs.
- Changed SQL connection string environment variable.
- Updated database context to use SQL Server.
- Introduced MockDataGenerator for generating mock data.
- Modified CoursesContext constructor and added OnConfiguring method.
- Implemented methods in MockDataGenerator for creating mock users, courses, and attendance records.

nicolasantolini committed:
Update dependencies and improve database configuration

- Added `local.settings.json` to `AttendanceSystem.Api.csproj`.
- Upgraded `Microsoft.ApplicationInsights` to version `2.23.0`.
- Updated `Microsoft.Azure.Functions.Worker` packages to version `2.0.1`.
- Upgraded `Microsoft.EntityFrameworkCore` and related packages to version `9.0.2`, including `Microsoft.EntityFrameworkCore.Design`.
- Changed connection string variable in `Program.cs` from `ConnectionString` to `SqlConnectionString` and updated database context configuration.
- Modified `MockDataGenerator.cs` to create courses with random teachers.
- Removed hardcoded SQL Server connection string from `CoursesContext.cs`.
- Updated migration files to reflect new database schema and relationships.
- Changed `SetAttendance` method in `Session.cs` to be asynchronous for improved performance.

joachimdekker commented on AttendanceSystem.Api/Program.cs:
```diff
@@ -33,6 +35,17 @@
+        userService,
+        courseService,
+        attendanceService
+    );
```
Please use the service provider for this. Add the `MockDataGenerator` to the service provider and use `scope.ServiceProvider.GetRequiredService<MockDataGenerator>()` for retrieving the instance. That way, we don't have to manually create objects (and the code gets cleaner).

joachimdekker commented on AttendanceSystem.Api/Program.cs:
```diff
@@ -33,6 +35,17 @@
+    );
+
+    //await MockDataGenerator.GenerateMockData();
+
```
I think we can use this call in development mode, since the database gets wiped after every run, we may want to have extra mock data in the database.

joachimdekker commented on AttendanceSystem.Data/CoursesContext.cs:
```diff
@@ -39,7 +39,14 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
+    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
+    {
+        base.OnConfiguring(optionsBuilder);
     }
```
Why are we overriding this method when we only call its base? Seems unnecessary.

joachimdekker commented on AttendanceSystem.Domain.Model/Session.cs:
```diff
@@ -38,8 +39,8 @@ public Session(Course course, List<StudentId> students, DateTime startTime, Date
 
-    public void SetAttendance(StudentId studentId, AttendanceKind kind)
-    {
+    public async Task SetAttendance(StudentId studentId, AttendanceKind kind)
```
There is no async call here, so why did you change the signature of this method? I think it can just be reverted to `void`.

joachimdekker reviewed (changes requested):
Very nice. Thanks for taking the time to upgrade the packages. I have left some comments.

nicolasantolini committed:
Refactor mock data generation and service registrations

- Updated `Program.cs` to use dependency injection for `MockDataGenerator` instead of manually creating service instances.
- Added `MockDataGenerator` registration in `Services.cs` to improve modularity and testability.
- Removed the `OnConfiguring` method in `CoursesContext.cs` for cleaner configuration management.
- Changed `SetAttendance` method in `Session.cs` from async to sync for improved performance.

joachimdekker committed:
Merge branch 'main' into MockDataGenerator

joachimdekker reviewed (approved):
LGTM
